hazelcast/src/main/java/com/hazelcast/util/executor/LoggingScheduledExecutor.java

```diff
@@ -50,28 +50,49 @@
  */
 public class LoggingScheduledExecutor extends ScheduledThreadPoolExecutor {
 
+    private static final int DEFAULT_PURGE_INITIAL_DELAY = 0;
+    private static final int DEFAULT_PURGE_PERIOD = 10;
+    private static final TimeUnit DEFAULT_PURGE_TIME_UNIT = TimeUnit.SECONDS;
+
     private final ILogger logger;
     private volatile boolean shutdownInitiated;
 
     public LoggingScheduledExecutor(ILogger logger, int corePoolSize, ThreadFactory threadFactory) {
         super(corePoolSize, threadFactory);
         this.logger = checkNotNull(logger, "logger cannot be null");
+        enablePeriodicPurge();
     }
 
     public LoggingScheduledExecutor(ILogger logger, int corePoolSize, ThreadFactory threadFactory,
                                     RejectedExecutionHandler handler) {
         super(corePoolSize, threadFactory, handler);
         this.logger = checkNotNull(logger, "logger cannot be null");
+        enablePeriodicPurge();
     }
 
     @Override
     protected <V> RunnableScheduledFuture<V> decorateTask(Runnable runnable, RunnableScheduledFuture<V> task) {
-        return new LoggingDelegatingFuture<V>(runnable, task, this);
+        return new LoggingDelegatingFuture<V>(runnable, task);
     }
 
     @Override
     protected <V> RunnableScheduledFuture<V> decorateTask(Callable<V> callable, RunnableScheduledFuture<V> task) {
-        return new LoggingDelegatingFuture<V>(callable, task, this);
+        return new LoggingDelegatingFuture<V>(callable, task);
+    }
+
+    /**
+     * {@link ScheduledThreadPoolExecutor#removeOnCancel policy is not working when tasks are decorated, causing task leaks in
+     * the workQueue. On the other hand, {@link ScheduledThreadPoolExecutor#remove(Runnable)} is an O(N) operation,
+     * not performaing well, causing severe operation delays. This periodic task will purge cancelled tasks from the queues,
+     * releasing effectively resources.
+     */
+    private void enablePeriodicPurge() {
```

Comment:
If `LoggingScheduledExecutor ` is created with just 1 thread, as it is now, this task can block that scheduler thread significant amount of time, hence causes pending scheduled tasks to delay.

Reply:
Instead you can schedule purge in `ExecutionService`:
```java
    public ExecutionServiceImpl(NodeEngineImpl nodeEngine) {
        this.nodeEngine = nodeEngine;
        [...]
        scheduleWithRepetition(new Runnable() {
            @Override
            public void run() {
                scheduledExecutorService.purge();
            }
        }, 1, 10, TimeUnit.SECONDS);
    }
```

Reply:
Fixed.